Send all jobs to planner

dtaniwaki opened the pull request:
Currently, we only pass limited number of jobs to planners, but some planners may want to evaluate all the jobs by themselves.

Some related topic has been discussed in https://github.com/retz/retz/issues/157 but we haven't concluded yet. My workaround here is let users configure it by the server configuration.

I didn't make the config `on/off` flag because we may want to add other strategy in the future.

Could you consider merging this change?

kuenishi commented on retz-server/src/main/java/io/github/retz/scheduler/ServerConfiguration.java:
```diff
@@ -87,6 +87,28 @@
+        Fit("fit"),
+        All("all");
+        static JobQueueStrategy getStrategy(String s) {
+            switch(s.toLowerCase()) {
```
This line introduces new warning in checkstyle. Checkstyle is introduced recently. 
```
[ant:checkstyle] [WARN] /home/kuenishi/src/retz/retz-server/src/main/java/io/github/retz/scheduler/ServerConfiguration.java:96:19: 'switch' の後にホワイトスペースがありません。 [WhitespaceAround]   
```

kuenishi commented on retz-server/src/main/java/io/github/retz/scheduler/ServerConfiguration.java:
```diff
@@ -241,6 +268,8 @@ public String getPlannerName() {
         return properties.getProperty(PLANNER_NAME, DEFAULT_PLANNER_NAME);
     }
 
+    public JobQueueStrategy getJobQueueStrategy() { return JobQueueStrategy.getStrategy(properties.getProperty(JOB_QUEUE_STRATEGY, DEFAULT_JOB_QUEUE_STRATEGY)); }
```
Ditto.
```
[ant:checkstyle] [WARN] /home/kuenishi/src/retz/retz-server/src/main/java/io/github/retz/scheduler/ServerConfiguration.java:271:51: 列 51の '{' の後には改行を入れてください。 [LeftCurly]  
```

kuenishi commented on retz-server/src/main/java/io/github/retz/scheduler/ServerConfiguration.java:
```diff
@@ -87,6 +87,28 @@
+    private static final String DEFAULT_JOB_QUEUE_STRATEGY = "all";
+    public enum JobQueueStrategy {
+        Fit("fit"),
+        All("all");
```
I would prefer all capital letters for enum names.

kuenishi commented on retz-server/src/main/java/io/github/retz/scheduler/RetzScheduler.java:
```diff
@@ -226,16 +226,28 @@ public void resourceOffers(SchedulerDriver driver, List<Protos.Offer> offers) {
+                    LOG.debug("found {} jobs", jobs.size());
+                    break;
+                default:
+                    LOG.debug("skip unknown strategy");
```
Considering startup check at `ServerConfiguration` constructor this must not happen; `AssertionError` is usually used in this software.

kuenishi commented on retz-server/src/main/java/io/github/retz/db/Database.java:
```diff
@@ -544,36 +544,44 @@ private void deleteApplication(Connection conn, String appid) throws SQLExceptio
+        int totalCpu = 0;
+        int totalMem = 0;
+
+        for (Job job: findAll(orderBy)) {
```
With this style, regardless of the strategy 'all' or 'fit' the scheduler must fetch and iterate all over the jobs in the database, because in `findAll` it fetches all jobs into the database. Although whether current code iterates over all jobs in the database or not depends on implementation of JDBC drivers, I don't want iteration over all jobs explicitly happen.

I wish if Java had supported `yield` in Python, it have been much simpler but I don't have clean answer for this. Maybe 'all' strategy sets cpu and totalMem as `Long.MAX_VALUE` or add another if condition in while clause.

dtaniwaki commented:
@kuenishi I implemented it almost from scratch based on your feedback. Could you check it again?
I added an option for `ALL` type of job queue and let it limit the number of jobs taken from the DB and don't touch the current implementation of `findFit`. In the future, we may add some option to `FIT` job queue type; e.g. `coefficient`

dtaniwaki commented:
I set the default job queue type to `fit` so other users won't be affected by this change.
